src/test/java/com/example/demo/WebAndSecurity/LoginValidationTest.java

```diff
@@ -0,0 +1,125 @@
+        String requestBody = TestUtils.asJsonString(new AuthRequest("ThisUsernameIsReeeeeeeeeeeeeeeeeaaaallyLong", "testPassword!"));
+
+        MvcResult result = mvc.perform(
+                        MockMvcRequestBuilders.post("/auth/login")
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .content(requestBody))
+                .andReturn();
+        assertThat(result.getResponse().getStatus()).isEqualTo(HttpStatus.BAD_REQUEST.value());
+        assertThat(result.getResponse().getContentAsString()).contains("length must be between");
+        assertThat(result.getResponse().getContentAsString()).contains("username");
+    }
+
+    @Test
+    public void ifShouldAlertIfPasswordTooShort() throws Exception {
+        String requestBody = TestUtils.asJsonString(new AuthRequest("testUsername", "aA!1"));
+
+        MvcResult result = mvc.perform(
+                        MockMvcRequestBuilders.post("/auth/login")
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .content(requestBody))
+                .andReturn();
+        assertThat(result.getResponse().getStatus()).isEqualTo(HttpStatus.BAD_REQUEST.value());
+        assertThat(result.getResponse().getContentAsString()).contains("Password should be");
+        assertThat(result.getResponse().getContentAsString()).contains("password");
+    }
+
+    @Test
+    public void ifShouldAlertIfPasswordTooLong() throws Exception {
+        String requestBody = TestUtils.asJsonString(new AuthRequest("testUsername", "aA!1eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee"));
+
+        MvcResult result = mvc.perform(
+                        MockMvcRequestBuilders.post("/auth/login")
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .content(requestBody))
+                .andReturn();
+        assertThat(result.getResponse().getStatus()).isEqualTo(HttpStatus.BAD_REQUEST.value());
+        assertThat(result.getResponse().getContentAsString()).contains("Password should be");
+        assertThat(result.getResponse().getContentAsString()).contains("password");
+    }
+
+    @Test
+    public void ifShouldAlertIfPasswordMissingUpperCase() throws Exception {
+        String requestBody = TestUtils.asJsonString(new AuthRequest("testUsername", "almostagoodpassword1!"));
+
+        MvcResult result = mvc.perform(
+                        MockMvcRequestBuilders.post("/auth/login")
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .content(requestBody))
+                .andReturn();
+        assertThat(result.getResponse().getStatus()).isEqualTo(HttpStatus.BAD_REQUEST.value());
+        assertThat(result.getResponse().getContentAsString()).contains("Password should be");
+        assertThat(result.getResponse().getContentAsString()).contains("password");
+    }
+
+    // etc etc
```

Comment:
Nitpick - maybe delete this comment just so the code on main is a bit cleaner 👍

Reply:
this is to indicate that for an actual production bit of code I'd do a bunch more.